tests/solana_tests/optimizations.rs

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    borsh_encoding::{visit_mut, VisitorMut},
+    BorshToken, VirtualMachineBuilder,
+};
+use once_cell::sync::Lazy;
+use rayon::prelude::*;
+use serde::Deserialize;
+use solang::codegen::Options;
+use std::{
+    env::var,
+    fs::{read_dir, read_to_string, File},
+    io::BufReader,
+    path::Path,
+};
+
+#[derive(Debug, Deserialize)]
+struct Calls {
+    constructor: Vec<BorshToken>,
+    function: Vec<(String, Vec<BorshToken>)>,
+}
+
+static NO_OPTIMIZATIONS: Lazy<Options> = Lazy::new(|| Options {
+    dead_storage: false,
+    constant_folding: false,
+    strength_reduce: false,
+    vector_to_slice: false,
+    common_subexpression_elimination: false,
+    ..Default::default()
+});
+
+#[test]
+fn optimizations() {
+    let calls = Path::new("tests/optimization_testcases/calls");
+
+    if let Ok(testname) = var("TESTNAME") {
+        run_test(&calls.join(testname).with_extension("json"));
+    } else {
+        let tests = read_dir(calls)
+            .unwrap()
+            .map(|entry| entry.unwrap().path())
+            .collect::<Vec<_>>();
+        tests.into_par_iter().for_each(|path| run_test(&path));
+    }
+}
+
+fn run_test(path: &Path) {
+    let file_stem = path.file_stem().unwrap();
+
+    // Known problematic test.
+    if file_stem == "b6339ad75e9175a6bf332a2881001b6c928734e2" {
+        return;
+    }
+
+    println!("testcase: {:?}", file_stem);
+
+    let file =
+        File::open(path).unwrap_or_else(|error| panic!("failed to open {path:?}: {error:?}"));
+    let reader = BufReader::new(file);
+    let calls: Calls = serde_json::from_reader(reader).unwrap();
+
+    let path = Path::new("tests/optimization_testcases/programs")
+        .join(file_stem)
+        .with_extension("sol");
+    let program = read_to_string(path).unwrap();
+
+    run_test_with_opts(
+        &program,
+        &calls,
+        [Options::default(), NO_OPTIMIZATIONS.clone()],
```

Comment:
Would be good to have a config corresponding to what `compile --release` uses; I don't think `Options::default()` is necessarily the same.

Reply:
Perhaps that could be a separate PR?